gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/auth/GcsDelegationTokens.java

```diff
@@ -238,6 +247,24 @@ public boolean isBoundToDT() {
       return getBoundDT();
     }
 
+    // not bounded tokens, but using shared token instantiation strategy.
+    // in this case we create or reuse token per renewer but share across multiple services.
+    if (this.tokenInstantiationStrategy == GCSTokenInstantiationStrategy.SHARED) {
+      Optional<String> renewerKey = Optional.ofNullable(renewer);
```

Comment:
Not sure if it's good idea to use optional as a map key, can we use some constant for empty values instead?

Reply:
Any string value can come from the client - I wanted to make sure that those values are not mixed with each other and tokens are not reused in the scenarios they are not supposed to.

Reply:
We can use an empty string if it's not a valid use input, but if we want to be more clear about uniqueness of shared token strategy we may want to use `AtomicReference` for shred token, instead of trying to put it in the map with non-shared tokens.

Reply:
I'm not confident that we can use an empty string as a sentinel value, considering wider Hadoop usage where only `null` represents a token without a renewer.

Another option would be to define a custom map key class that carefully handles the semantics of null vs. non-null renewer, but that seemed like the heaviest option compared to alternatives.

I'm unclear about the `AtomicReference` suggestion. Is the idea that a token with `null` renewer goes into separate storage from the map? Can you please elaborate?

Reply:
@medb another option I was considering adding an additional field for storing token when renewer is `null`, but it seems to be excessive.